### DISCUSSION TITLE
[Bug]: [tests] `within(canvasElement)` cannot find elements rendered via portal to the document.body

### DISCUSSION THREAD

**Discussion by csantos1113** [storybookjs/storybook#27024] [Help]
### Describe the bug

https://github.com/storybookjs/storybook/tree/next/code/lib/test says I should be using `within(canvasElement)` in my play functions, instead of `screen`. but it doesn't work for portal elements rendered to the document.body

### To Reproduce

- Go to https://stackblitz.com/edit/github-t5ssay?file=src%2Fstories%2FPage.stories.ts
- Notice that `screen.getByText` works, but shows a warning in the console saying that I should use `within(canvasElement)` instead
- Notice that `canvas.getByText` fails because the element is rendered via portal to the document.body

### System

```bash
Storybook Environment Info:

  System:
    OS: Linux 5.0 undefined
    CPU: (8) x64 Intel(R) Core(TM) i9-9880H CPU @ 2.30GHz
    Shell: 1.0 - /bin/jsh
  Binaries:
    Node: 18.18.0 - /usr/local/bin/node
    Yarn: 1.22.19 - /usr/local/bin/yarn
    npm: 10.2.3 - /usr/local/bin/npm <----- active
    pnpm: 8.15.6 - /usr/local/bin/pnpm
  npmPackages:
    @storybook/addon-essentials: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/addon-interactions: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/addon-links: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/addon-onboarding: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/addon-webpack5-compiler-swc: ^1.0.2 => 1.0.2 
    @storybook/blocks: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/react: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/react-webpack5: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    @storybook/test: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
    storybook: ^8.1.0-alpha.7 => 8.1.0-alpha.7 
```

### Additional context

Real live case: I'm opening a Dialog/Drawer which is rendered via portals.

Any recommendation on how to reach elements rendered via portals to the document.body that adheres to the expected behavior defined by Storybook?

**Comment by valentinpalkovic** [comment]
Instead of `within(canvasElement)` you could use `within(canvasElement.parentNode)` to target the body and therefore to assert your dialog.

**Comment by csantos1113** [comment]
> Instead of `within(canvasElement)` you could use `within(canvasElement.parentNode)` to target the body and therefore to assert your dialog.

Thanks for the response, but TypeScript doesn't seem to like it :(

<img width="985" alt="image" src="https://github.com/storybookjs/storybook/assets/87080313/032fec7c-4ff9-465d-a423-a61130d079cc">


**UPDATE:** ah, I'm using storybook 7.6.17 locally, which seems to have the type issue

And also, in 8.x I need to do `within(canvasElement.parentElement)` instead

**Comment by csantos1113** [comment]
I actually seems my types issue are not related to Storybook version. when I use `canvastElement.parentElement` I get `HTMLElement | null`. But not sure where those wrong types are coming from.

I'm using:

```
    "typescript": "5.4.2",
    "@types/react": "17.0.50",
    "@types/react-dom": "17.0.17",
```

**Comment by csantos1113** [comment]
UPDATE: 

@valentinpalkovic - Okay, I'm not sure where did I get that other stackblitz example from; but here is an up-to-date as of today: https://stackblitz.com/edit/github-yrt7dh?file=src%2Fstories%2FPage.stories.ts&preset=node

Which is failing the same TypeScript error I'm experiencing locally

<img width="992" alt="image" src="https://github.com/storybookjs/storybook/assets/87080313/875cc2fe-a229-4720-846a-450e499f7762">

**Comment by przemyslaw-wlodek** [comment]
@csantos1113 the error isn't coming from any of the libraries you mentioned but from the nature of DOM being untyped. TypeScript can't guarantee that your `HTMLElement` (`canvasElement`) has a `parentNode`.

I see 2 possible options here:
1. Quick and a little bit dirty [non-null type assertion](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator) (`within(canvasElement.parentNode!)`) since it's only test + you are sure that there's always some `body` above your element.
2. Create a custom Page decorator that renders the element that could be used as a portal container (it will be a part of canvas then). When you have a `ref` to your custom portal container, then you can easily pass it to your Portal component (which defaults to `document.body`).

I had a very similar problem when testing [Radix](https://www.radix-ui.com/) components that use portals. You may find some inspiration in [this PR](https://github.com/input-output-hk/lace/pull/1109).

**Comment by valentinpalkovic** [comment]
I will close this issue since a solution was provided and the Typescript issue was elaborated above

**Comment by mike-wheel** [comment]
For anyone else (or future me) looking for another solution, I've found that this works well:
```typescript
const body = within(canvasElement.ownerDocument.body);
```

**Reply by RNm-dove** [reply]
Wonderful! It works perfectly for me.

**Comment by Roman86** [comment]
also, even if you aren't using a body, but still have an issue with getBy methods (getting "Unable to find an element" when it tries to access an element before it's actually rendered, usually noticeable when you reload a story page - the play test fails, but if you navigate between stories - it passes) - try asynchronous findBy. Worked for me.
https://github.com/storybookjs/testing-library/issues/25#issuecomment-1184255825
